README.md
# azure_inventory

#### Table of Contents

1. [Description](#description)
2. [Requirements](#requirements)
3. [Usage](#usage)

## Description

This module includes a Bolt plugin to generate Bolt targets from Azure VMs.

## Requirements

You will need a client ID and secret in order to authenticate with Azure. The simplest way to generate these is to use the [Azure CLI tool](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli).

Run `az ad sp create-for-rbac` and copy the `appId` and `password` fields. These correspond to the `client_id` and `client_secret` parameters to the task.

Alternatively, you can follow the instructions to [register a client application with Azure AD](https://docs.microsoft.com/en-us/rest/api/azure/#register-your-client-application-with-azure-ad).

You will also need to know your subscription ID and tenant ID. If you're using the Azure CLI, you can retrieve these with `az account show`. These will be in the `id` and `tenantId` fields respectively.

## Usage

The plugin supports looking up virtual machines and virtual machine scale sets. It supports several fields:

- `resource_group`: The resource group to filter by (optional)
- `scale_set`: The scale set to filter by (optional, requires that `resource_group` also be set)
- `location`: The location to filter by (optional)
- `tags`: A list of one or more tags to filter by (optional, tags are `name: value` pairs and instances must match all listed tags)

Accessing Azure resources requires credentials for signing all API requests. Each credential will be looked up from an environment variable if not set in the inventory config.

- `client_id`: The Azure client application ID (defaults to `$AZURE_CLIENT_ID`)
- `client_secret`: The Azure client application secret (defaults to `$AZURE_CLIENT_SECRET`)
- `tenant_id`: The Azure AD tenant ID (defaults to `$AZURE_TENANT_ID`)
- `subscription_id`: The Azure subscription ID (defaults to `$AZURE_SUBSCRIPTION_ID`)

Bolt will only target virtual machines and virtual machine scale sets that have a public IP address. The `uri` of the target will be set to the public IP address and the `name` will be set to either the fully qualified domain name if one exists or the instance name otherwise.

If `scale_set` is not provided, Bolt will not find VMs that are defined by a scale set.

### Examples

```yaml
# inventory.yaml
version: 2
groups:
  - name: azure-vms
    targets:
      - _plugin: azure_inventory
        tenant_id: xxxx-xxx-xxxx
        client_id: xxxx-xxx-xxxx
        client_secret: xxxx-xxx-xxxx
        subscription_id: xxxx-xxx-xxxx
        location: eastus
        resource_group: bolt
        tags:
          foo: bar
          baz: bak
  - name: azure-scale-sets
    targets:
      - _plugin: azure_inventory
        tenant_id: xxxx-xxx-xxxx
        client_id: xxxx-xxx-xxxx
        client_secret: xxxx-xxx-xxxx
        subscription_id: xxxx-xxx-xxxx
        location: eastus2
        resource_group: puppet
        scale_set: bolt
        tags:
          foo: bar
```


CHANGELOG.md
## Release 0.5.0

### New features

* **Bump maximum Puppet version to include 7.x** ([#14](https://github.com/puppetlabs/puppetlabs-azure_inventory/pull/14))

## Release 0.4.1

### Bug fixes

* **Add PDK as a gem dependency**

  PDK is now a gem dependency for the module release pipeline

## Release 0.4.0

### New features

* **Add debugging statements to task errors**
  ([#9](https://github.com/puppetlabs/puppetlabs-azure_inventory/pull/9))

  Error objects returned from the `resolve_reference` task now includes
  debugging statements that describe the steps the task is taking under
  the `details` key.

### Bug fixes

* **Add missing dependency to module metadata**
  ([#10](https://github.com/puppetlabs/puppetlabs-azure_inventory/pull/9))

  The module metadata now includes `ruby_plugin_helper` and `ruby_task_helper`
  as dependencies.

## Release 0.3.0

### New features

* **Set `resolve_reference` task to private** ([#6](https://github.com/puppetlabs/puppetlabs-azure_inventory/pull/6))

    The `resolve_reference` task has been set to `private` so it no longer appears in UI lists.

## Release 0.2.0

**Changes**

This converts the module to a Bolt plugin, which includes renaming the `inventory_targets` task to `resolve_references`.

## Release 0.1.0

This is the initial release.


spec/tasks/inventory_targets_spec.rb
# frozen_string_literal: true

require 'spec_helper'
require_relative '../../tasks/inventory_targets.rb'

describe AzureInventory do
  def with_env(key, val)
    old_val = ENV[key]
    ENV[key] = val
    yield
  ensure
    ENV[key] = old_val
  end

  def fixture(name)
    JSON.parse(File.read(File.join(__dir__, "../fixtures/responses/#{name}.json")))
  end

  let(:tenant_id) { SecureRandom.uuid }
  let(:client_id) { SecureRandom.uuid }
  let(:client_secret) { SecureRandom.uuid }
  let(:subscription_id) { SecureRandom.uuid }
  let(:opts) do
    { tenant_id: tenant_id,
      client_id: client_id,
      client_secret: client_secret,
      subscription_id: subscription_id }
  end

  before :each do
    # Make sure we don't make any HTTP requests
    allow(subject).to receive(:request).and_return(nil)
  end

  describe "#inventory_targets" do
    before :each do
      allow(subject).to receive(:vms).and_return(fixture('vms'))
      allow(subject).to receive(:ip_addresses).and_return(fixture('ip_addresses'))
      allow(subject).to receive(:nics).and_return(fixture('nics'))
    end

    it 'generates a list of targets with ip addresses' do
      allow(subject).to receive(:token).and_return({})

      targets = [
        { "name" => "rgtest-1", "uri" => "52.160.41.155" },
        { "name" => "test-instance-1", "uri" => "40.118.207.76" }
      ]
      expect(subject.inventory_targets(opts)).to eq(targets)
    end
  end

  describe "#credentials" do
    it 'reads credentials from the supplied options' do
      expect(subject.credentials(opts)).to eq(
        'tenant_id' => tenant_id,
        'client_id' => client_id,
        'client_secret' => client_secret,
        'subscription_id' => subscription_id
      )
    end

    %i[tenant_id client_id client_secret subscription_id].each do |key|
      it "accepts #{key} from the environment" do
        env_key = "AZURE_#{key.upcase}"
        val = opts.delete(key)

        with_env(env_key, val) do
          expect(subject.credentials(opts)).to eq(
            'tenant_id' => tenant_id,
            'client_id' => client_id,
            'client_secret' => client_secret,
            'subscription_id' => subscription_id
          )
        end
      end

      it "fails if #{key} is unset" do
        opts.delete(key)
        expect { subject.credentials(opts) }.to raise_error(TaskHelper::Error, /#{key} must be specified/)
      end
    end
  end

  describe "#get_all_results" do
    let(:token) { { 'token_type' => 'foo', 'access_token' => 'bar' } }

    it 'loops through each page of results' do
      [1, 2, 3].each do |i|
        uri = URI.parse("https://example.com/page/#{i}")
        response = { 'value' => ["#{i}a", "#{i}b", "#{i}c"],
                     'nextLink' => "https://example.com/page/#{i + 1}" }
        allow(subject).to receive(:request).with(:Get, uri, nil, anything).and_return(response)
      end
      uri = URI.parse("https://example.com/page/4")
      response = { 'value' => %w[4a 4b 4c] }
      allow(subject).to receive(:request).with(:Get, uri, nil, anything).and_return(response)

      results = subject.get_all_results("https://example.com/page/1", token)
      expect(results).to eq(%w[1a 1b 1c 2a 2b 2c 3a 3b 3c 4a 4b 4c])
    end
  end

  describe "#task" do
    it 'fails if scale_set is requested without resource_group' do
      result = subject.task(scale_set: 'something')
      expect(result).to have_key(:_error)
      expect(result[:_error]['msg']).to match(/resource_group must be specified in order to filter by scale_set/)
    end

    it 'returns the list of targets' do
      targets = [
        { "uri": "1.2.3.4", "name": "my-instance" },
        { "uri": "1.2.3.5", "name": "my-other-instance" }
      ]
      allow(subject).to receive(:inventory_targets).and_return(targets)

      result = subject.task(opts)
      expect(result).to have_key(:targets)
      expect(result[:targets]).to eq(targets)
    end

    it 'returns an error if one is raised' do
      error = TaskHelper::Error.new('something went wrong', 'bolt.test/error')
      allow(subject).to receive(:inventory_targets).and_raise(error)
      result = subject.task({})

      expect(result).to have_key(:_error)
      expect(result[:_error]['msg']).to match(/something went wrong/)
    end
  end
end


spec/tasks/resolve_reference_spec.rb
# frozen_string_literal: true

require 'spec_helper'
require_relative '../../tasks/resolve_reference.rb'

describe AzureInventory do
  def with_env(key, val)
    old_val = ENV[key]
    ENV[key] = val
    yield
  ensure
    ENV[key] = old_val
  end

  def fixture(name)
    JSON.parse(File.read(File.join(__dir__, "../fixtures/responses/#{name}.json")))
  end

  let(:tenant_id) { SecureRandom.uuid }
  let(:client_id) { SecureRandom.uuid }
  let(:client_secret) { SecureRandom.uuid }
  let(:subscription_id) { SecureRandom.uuid }
  let(:opts) do
    { tenant_id: tenant_id,
      client_id: client_id,
      client_secret: client_secret,
      subscription_id: subscription_id }
  end

  before :each do
    # Make sure we don't make any HTTP requests
    allow(subject).to receive(:request).and_return(nil)
  end

  describe "#resolve_reference" do
    before :each do
      allow(subject).to receive(:vms).and_return(fixture('vms'))
      allow(subject).to receive(:ip_addresses).and_return(fixture('ip_addresses'))
      allow(subject).to receive(:nics).and_return(fixture('nics'))
    end

    it 'generates a list of targets with ip addresses' do
      allow(subject).to receive(:token).and_return({})

      targets = [
        { "name" => "rgtest-1", "uri" => "52.160.41.155" },
        { "name" => "test-instance-1", "uri" => "40.118.207.76" }
      ]
      expect(subject.resolve_reference(opts)).to eq(targets)
    end
  end

  describe "#credentials" do
    it 'reads credentials from the supplied options' do
      expect(subject.credentials(opts)).to eq(
        'tenant_id' => tenant_id,
        'client_id' => client_id,
        'client_secret' => client_secret,
        'subscription_id' => subscription_id
      )
    end

    %i[tenant_id client_id client_secret subscription_id].each do |key|
      it "accepts #{key} from the environment" do
        env_key = "AZURE_#{key.upcase}"
        val = opts.delete(key)

        with_env(env_key, val) do
          expect(subject.credentials(opts)).to eq(
            'tenant_id' => tenant_id,
            'client_id' => client_id,
            'client_secret' => client_secret,
            'subscription_id' => subscription_id
          )
        end
      end

      it "fails if #{key} is unset" do
        opts.delete(key)
        expect { subject.credentials(opts) }.to raise_error(TaskHelper::Error, /#{key} must be specified/)
      end
    end
  end

  describe "#get_all_results" do
    let(:token) { { 'token_type' => 'foo', 'access_token' => 'bar' } }

    it 'loops through each page of results' do
      [1, 2, 3].each do |i|
        uri = URI.parse("https://example.com/page/#{i}")
        response = { 'value' => ["#{i}a", "#{i}b", "#{i}c"],
                     'nextLink' => "https://example.com/page/#{i + 1}" }
        allow(subject).to receive(:request).with(:Get, uri, nil, anything).and_return(response)
      end
      uri = URI.parse("https://example.com/page/4")
      response = { 'value' => %w[4a 4b 4c] }
      allow(subject).to receive(:request).with(:Get, uri, nil, anything).and_return(response)

      results = subject.get_all_results("https://example.com/page/1", token)
      expect(results).to eq(%w[1a 1b 1c 2a 2b 2c 3a 3b 3c 4a 4b 4c])
    end
  end

  describe "#task" do
    it 'fails if scale_set is requested without resource_group' do
      result = subject.task(scale_set: 'something')
      expect(result).to have_key(:_error)
      expect(result[:_error]['msg']).to match(/resource_group must be specified in order to filter by scale_set/)
    end

    it 'returns the list of targets' do
      targets = [
        { "uri": "1.2.3.4", "name": "my-instance" },
        { "uri": "1.2.3.5", "name": "my-other-instance" }
      ]
      allow(subject).to receive(:resolve_reference).and_return(targets)

      result = subject.task(opts)
      expect(result).to have_key(:value)
      expect(result[:value]).to eq(targets)
    end

    it 'returns an error if one is raised' do
      error = TaskHelper::Error.new('something went wrong', 'bolt.test/error')
      allow(subject).to receive(:resolve_reference).and_raise(error)
      result = subject.task({})

      expect(result).to have_key(:_error)
      expect(result[:_error]['msg']).to match(/something went wrong/)
    end
  end
end


tasks/inventory_targets.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '../../ruby_task_helper/files/task_helper.rb'
require 'net/http'
require 'openssl'

class AzureInventory < TaskHelper
  def inventory_targets(opts)
    creds = credentials(opts)
    token = token(creds)
    vms = nics = ips = nil
    threads = []
    threads << Thread.start { vms = vms(token, creds, opts) }
    threads << Thread.start { nics = index_by_id(nics(token, creds, opts)) }
    threads << Thread.start { ips = index_by_id(ip_addresses(token, creds, opts)) }
    threads.each(&:join)

    vms.map do |vm|
      interfaces = vm.dig('properties', 'networkProfile', 'networkInterfaces')
      nic_ids = interfaces.partition { |nic| nic['primary'] }.flatten.map { |nic| nic['id'] }
      vm_nics = nics.values_at(*nic_ids)

      ip_configs = vm_nics.flat_map do |nic|
        nic.dig('properties', 'ipConfigurations') || nic.dig('properties', 'ipConfiguration')
      end
      ip_ids = ip_configs.map { |ip| ip.dig('properties', 'publicIPAddress', 'id') }
      vm_ips = ips.values_at(*ip_ids)

      ip = vm_ips.compact.first
      next unless ip
      {
        'name' => ip.dig('properties', 'dnsSettings', 'fqdn') || vm['name'],
        'uri' => ip.dig('properties', 'ipAddress')
      }.compact
    end.compact
  end

  # Hash of required credentials for authorizing with the Azure REST API
  # These values can be set in 2 locations - inventory config or environment variables
  def credentials(opts)
    creds = {
      'tenant_id' => (opts[:tenant_id] || ENV['AZURE_TENANT_ID']),
      'client_id' => (opts[:client_id] || ENV['AZURE_CLIENT_ID']),
      'client_secret' => (opts[:client_secret] || ENV['AZURE_CLIENT_SECRET']),
      'subscription_id' => (opts[:subscription_id] || ENV['AZURE_SUBSCRIPTION_ID'])
    }

    missing_keys = creds.select { |_k, v| v.nil? }.keys
    if missing_keys.any?
      msg = "Parameters #{missing_keys.join(', ')} must be specified or set as environment variables"
      raise TaskHelper::Error.new(msg, 'bolt-plugin/validation-error')
    end

    creds
  end

  def get_all_results(url, token)
    header = {
      'Authorization' => "#{token['token_type']} #{token['access_token']}"
    }

    instances = []

    while url
      # Update the URI and make the next request
      uri = URI.parse(url)
      result = request(:Get, uri, nil, header)

      # Add the VMs to the list of instances
      instances.concat(result['value'])

      # Continue making requests until there is no longer a nextLink
      url = result['nextLink']
    end

    instances
  end

  # Requests for VMs and scale sets are on a per-subscription basis
  # You can also request VMs by a specific resource group
  # Scale sets are always requested by resource group and scale set name
  #
  # Since each request only returns up to 1,000 results, requests will continue to be
  # sent until there is no longer a nextLink token in the result set
  def ip_addresses(token, creds, opts)
    # XXX What happens if I have multiple IP addresses for a single host?
    url = if opts[:resource_group]
            if opts[:scale_set]
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
              "publicIPAddresses?api-version=2017-03-30"
            else
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Network/" \
              "publicIPAddresses?api-version=2019-07-01"
            end
          else
            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
            "providers/Microsoft.Network/publicIPAddresses?api-version=2019-07-01"
          end

    get_all_results(url, token)
  end

  def vms(token, creds, opts)
    url = if opts[:resource_group]
            if opts[:scale_set]
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
              "virtualmachines?api-version=2017-03-30"
            else
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualmachines?api-version=2019-07-01"
            end
          else
            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
            "providers/Microsoft.Compute/virtualmachines?api-version=2019-07-01"
          end

    vms = get_all_results(url, token)

    # Filter by location
    if opts[:location]
      vms.select! { |vm| vm['location'] == opts[:location] }
    end

    # Filter by tags - tags are ANDed
    if opts[:tags]
      # Tag names are case insensitive, values are case sensitive
      expected_tags = opts[:tags].map { |name, value| [name.to_s.downcase, value] }.to_h

      vms.select! do |vm|
        present_tags = vm.fetch('tags', {}).map { |name, value| [name.downcase, value] }.to_h
        # Hash <= checks whether the lhs is a subset of the rhs
        expected_tags <= present_tags
      end
    end

    vms
  end

  def nics(token, creds, opts)
    url = if opts[:resource_group]
            if opts[:scale_set]
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
              "networkinterfaces?api-version=2017-03-30"
            else
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Network/" \
              "networkInterfaces?api-version=2019-07-01"
            end
          else
            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
            "providers/Microsoft.Network/networkInterfaces?api-version=2019-07-01"
          end

    get_all_results(url, token)
  end

  def index_by_id(objects)
    objects.each_with_object({}) do |obj, acc|
      acc[obj['id']] = obj
    end
  end

  # Uses the client credentials grant flow
  # https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-oauth2-client-creds-grant-flow
  def token(creds)
    data = {
      grant_type: 'client_credentials',
      client_id: creds['client_id'],
      client_secret: creds['client_secret'],
      resource: 'https://management.azure.com'
    }

    uri = URI.parse("https://login.microsoftonline.com/#{creds['tenant_id']}/oauth2/token")

    request(:Post, uri, data)
  end

  def request(verb, uri, data, header = {})
    # Create the client
    client = Net::HTTP.new(uri.host, uri.port)

    # Azure REST API always uses SSL
    client.use_ssl = true
    client.verify_mode = OpenSSL::SSL::VERIFY_PEER

    # Build the request
    request = Net::HTTP.const_get(verb).new(uri.request_uri, header)

    # Build the query if there's data to send
    query = URI.encode_www_form(data) if data

    # Send the request
    begin
      response = client.request(request, query)
    rescue StandardError => e
      raise TaskHelper::Error.new(
        "Failed to connect to #{uri}: #{e.message}",
        'bolt.plugin/azure-http-error'
      )
    end

    case response
    when Net::HTTPOK
      JSON.parse(response.body)
    else
      result = JSON.parse(response.body)
      # Some responses have an error_description string and others have an
      # error object with a message string embedded.
      err = if result.key?('error_description')
              result['error_description']
            elsif result['error'].is_a?(Hash) && result['error'].key?('message')
              result['error']['message']
            else
              "Unknown error"
            end
      m = String.new("#{response.code} \"#{response.msg}\"")
      m += ": #{err}" if err
      raise TaskHelper::Error.new(m, 'bolt.plugin/azure-http-error')
    end
  end

  def task(opts)
    if opts[:scale_set] && !opts[:resource_group]
      msg = "resource_group must be specified in order to filter by scale_set"
      raise TaskHelper::Error.new(msg, 'bolt.plugin/validation-error')
    end

    targets = inventory_targets(opts)
    return { targets: targets }
  rescue TaskHelper::Error => e
    # ruby_task_helper doesn't print errors under the _error key, so we have to
    # handle that ourselves
    return { _error: e.to_h }
  end
end

if $PROGRAM_NAME == __FILE__
  AzureInventory.run
end


tasks/resolve_reference.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '../../ruby_task_helper/files/task_helper.rb'
require 'net/http'
require 'openssl'

class AzureInventory < TaskHelper
  def resolve_reference(opts)
    creds = credentials(opts)
    token = token(creds)
    vms = nics = ips = nil
    threads = []
    threads << Thread.start { vms = vms(token, creds, opts) }
    threads << Thread.start { nics = index_by_id(nics(token, creds, opts)) }
    threads << Thread.start { ips = index_by_id(ip_addresses(token, creds, opts)) }
    threads.each(&:join)

    vms.map do |vm|
      interfaces = vm.dig('properties', 'networkProfile', 'networkInterfaces')
      nic_ids = interfaces.partition { |nic| nic['primary'] }.flatten.map { |nic| nic['id'] }
      vm_nics = nics.values_at(*nic_ids)

      ip_configs = vm_nics.flat_map do |nic|
        nic.dig('properties', 'ipConfigurations') || nic.dig('properties', 'ipConfiguration')
      end
      ip_ids = ip_configs.map { |ip| ip.dig('properties', 'publicIPAddress', 'id') }
      vm_ips = ips.values_at(*ip_ids)

      ip = vm_ips.compact.first
      next unless ip
      {
        'name' => ip.dig('properties', 'dnsSettings', 'fqdn') || vm['name'],
        'uri' => ip.dig('properties', 'ipAddress')
      }.compact
    end.compact
  end

  # Hash of required credentials for authorizing with the Azure REST API
  # These values can be set in 2 locations - inventory config or environment variables
  def credentials(opts)
    debug("Gathering credentials")
    creds = {
      'tenant_id' => (opts[:tenant_id] || ENV['AZURE_TENANT_ID']),
      'client_id' => (opts[:client_id] || ENV['AZURE_CLIENT_ID']),
      'client_secret' => (opts[:client_secret] || ENV['AZURE_CLIENT_SECRET']),
      'subscription_id' => (opts[:subscription_id] || ENV['AZURE_SUBSCRIPTION_ID'])
    }

    missing_keys = creds.select { |_k, v| v.nil? }.keys
    if missing_keys.any?
      msg = "Parameters #{missing_keys.join(', ')} must be specified or set as environment variables"
      raise TaskHelper::Error.new(msg, 'bolt-plugin/validation-error', 'debug' => debug_statements)
    end

    creds
  end

  def get_all_results(url, token)
    header = {
      'Authorization' => "#{token['token_type']} #{token['access_token']}"
    }

    instances = []

    while url
      debug("Making request to #{url}")

      # Update the URI and make the next request
      uri = URI.parse(url)
      result = request(:Get, uri, nil, header)

      # Add the VMs to the list of instances
      instances.concat(result['value'])

      # Continue making requests until there is no longer a nextLink
      url = result['nextLink']
    end

    instances
  end

  # Requests for VMs and scale sets are on a per-subscription basis
  # You can also request VMs by a specific resource group
  # Scale sets are always requested by resource group and scale set name
  #
  # Since each request only returns up to 1,000 results, requests will continue to be
  # sent until there is no longer a nextLink token in the result set
  def ip_addresses(token, creds, opts)
    debug("Requesting data for IP addresses")

    # XXX What happens if I have multiple IP addresses for a single host?
    url = if opts[:resource_group]
            if opts[:scale_set]
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
              "publicIPAddresses?api-version=2017-03-30"
            else
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Network/" \
              "publicIPAddresses?api-version=2019-07-01"
            end
          else
            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
            "providers/Microsoft.Network/publicIPAddresses?api-version=2019-07-01"
          end

    get_all_results(url, token)
  end

  def vms(token, creds, opts)
    debug("Requesting data for virtual machines")

    url = if opts[:resource_group]
            if opts[:scale_set]
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
              "virtualmachines?api-version=2017-03-30"
            else
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualmachines?api-version=2019-07-01"
            end
          else
            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
            "providers/Microsoft.Compute/virtualmachines?api-version=2019-07-01"
          end

    vms = get_all_results(url, token)

    # Filter by location
    if opts[:location]
      vms.select! { |vm| vm['location'] == opts[:location] }
    end

    # Filter by tags - tags are ANDed
    if opts[:tags]
      # Tag names are case insensitive, values are case sensitive
      expected_tags = opts[:tags].map { |name, value| [name.to_s.downcase, value] }.to_h

      vms.select! do |vm|
        present_tags = vm.fetch('tags', {}).map { |name, value| [name.downcase, value] }.to_h
        # Hash <= checks whether the lhs is a subset of the rhs
        expected_tags <= present_tags
      end
    end

    vms
  end

  def nics(token, creds, opts)
    debug("Requesting data for network interfaces")

    url = if opts[:resource_group]
            if opts[:scale_set]
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Compute/" \
              "virtualMachineScaleSets/#{opts[:scale_set]}/" \
              "networkinterfaces?api-version=2017-03-30"
            else
              "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
              "resourceGroups/#{opts[:resource_group]}/providers/Microsoft.Network/" \
              "networkInterfaces?api-version=2019-07-01"
            end
          else
            "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
            "providers/Microsoft.Network/networkInterfaces?api-version=2019-07-01"
          end

    get_all_results(url, token)
  end

  def index_by_id(objects)
    objects.each_with_object({}) do |obj, acc|
      acc[obj['id']] = obj
    end
  end

  # Uses the client credentials grant flow
  # https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-oauth2-client-creds-grant-flow
  def token(creds)
    debug("Requesting authorization token")

    data = {
      grant_type: 'client_credentials',
      client_id: creds['client_id'],
      client_secret: creds['client_secret'],
      resource: 'https://management.azure.com'
    }

    uri = URI.parse("https://login.microsoftonline.com/#{creds['tenant_id']}/oauth2/token")

    request(:Post, uri, data)
  end

  def request(verb, uri, data, header = {})
    # Create the client
    client = Net::HTTP.new(uri.host, uri.port)

    # Azure REST API always uses SSL
    client.use_ssl = true
    client.verify_mode = OpenSSL::SSL::VERIFY_PEER

    # Build the request
    request = Net::HTTP.const_get(verb).new(uri.request_uri, header)

    # Build the query if there's data to send
    query = URI.encode_www_form(data) if data

    # Send the request
    begin
      response = client.request(request, query)
    rescue StandardError => e
      raise TaskHelper::Error.new(
        "Failed to connect to #{uri}: #{e.message}",
        'bolt.plugin/azure-http-error',
        'debug' => debug_statements
      )
    end

    case response
    when Net::HTTPOK
      JSON.parse(response.body)
    else
      result = JSON.parse(response.body)
      # Some responses have an error_description string and others have an
      # error object with a message string embedded.
      err = if result.key?('error_description')
              result['error_description']
            elsif result['error'].is_a?(Hash) && result['error'].key?('message')
              result['error']['message']
            else
              "Unknown error"
            end
      m = String.new("#{response.code} \"#{response.msg}\"")
      m += ": #{err}" if err
      raise TaskHelper::Error.new(m, 'bolt.plugin/azure-http-error', 'debug' => debug_statements)
    end
  end

  def task(opts)
    if opts[:scale_set] && !opts[:resource_group]
      msg = "resource_group must be specified in order to filter by scale_set"
      raise TaskHelper::Error.new(msg, 'bolt.plugin/validation-error')
    end

    targets = resolve_reference(opts)
    return { value: targets }
  rescue TaskHelper::Error => e
    # ruby_task_helper doesn't print errors under the _error key, so we have to
    # handle that ourselves
    return { _error: e.to_h }
  end
end

if $PROGRAM_NAME == __FILE__
  AzureInventory.run
end
